pkg/git/libgit2/checkout.go

```diff
@@ -72,177 +73,129 @@ type CheckoutBranch struct {
-		}
-		if opts.TransportOptionsURL == "" {
-			return nil, fmt.Errorf("can't use managed transport without a valid transport auth id.")
-		}
-		managed.AddTransportOptions(opts.TransportOptionsURL, managed.TransportOptions{
-			TargetURL:    url,
-			AuthOpts:     opts,
-			ProxyOptions: &git2go.ProxyOptions{Type: git2go.ProxyTypeAuto},
-			Context:      ctx,
-		})
-		url = opts.TransportOptionsURL
-		remoteCallBacks := managed.RemoteCallbacks()
-		defer managed.RemoveTransportOptions(opts.TransportOptionsURL)
-
-		repo, remote, err := initializeRepoWithRemote(ctx, path, url, opts)
-		if err != nil {
-			return nil, err
-		}
-		// Open remote connection.
-		err = remote.ConnectFetch(&remoteCallBacks, nil, nil)
-		if err != nil {
-			remote.Free()
-			repo.Free()
-			return nil, fmt.Errorf("unable to fetch-connect to remote '%s': %w", managed.EffectiveURL(url), gitutil.LibGit2Error(err))
-		}
-		defer func() {
-			remote.Disconnect()
-			remote.Free()
-			repo.Free()
-		}()
-
-		// When the last observed revision is set, check whether it is still the
-		// same at the remote branch. If so, short-circuit the clone operation here.
-		if c.LastRevision != "" {
-			heads, err := remote.Ls(c.Branch)
-			if err != nil {
-				return nil, fmt.Errorf("unable to remote ls for '%s': %w", managed.EffectiveURL(url), gitutil.LibGit2Error(err))
-			}
-			if len(heads) > 0 {
-				hash := heads[0].Id.String()
-				currentRevision := fmt.Sprintf("%s/%s", c.Branch, hash)
-				if currentRevision == c.LastRevision {
-					// Construct a partial commit with the existing information.
-					c := &git.Commit{
-						Hash:      git.Hash(hash),
-						Reference: "refs/heads/" + c.Branch,
-					}
-					return c, nil
-				}
-			}
-		}
-
-		// Limit the fetch operation to the specific branch, to decrease network usage.
-		err = remote.Fetch([]string{c.Branch},
-			&git2go.FetchOptions{
-				DownloadTags:    git2go.DownloadTagsNone,
-				RemoteCallbacks: remoteCallBacks,
-			},
-			"")
-		if err != nil {
-			return nil, fmt.Errorf("unable to fetch remote '%s': %w",
-				managed.EffectiveURL(url), gitutil.LibGit2Error(err))
-		}
-
-		branch, err := repo.References.Lookup(fmt.Sprintf("refs/remotes/origin/%s", c.Branch))
-		if err != nil {
-			return nil, fmt.Errorf("unable to lookup branch '%s' for '%s': %w",
-				c.Branch, managed.EffectiveURL(url), gitutil.LibGit2Error(err))
-		}
-		defer branch.Free()
+	// We store the target URL and auth options mapped to a unique ID. We overwrite the target URL
+	// with the TransportOptionsURL, because managed transports don't provide a way for any kind of
+	// dependency injection. This lets us have a way of doing interop between application level code
+	// and transport level code.
+	// Performing all fetch operations with the TransportOptionsURL as the URL, lets the managed
+	// transport action use it to fetch the registered transport options which contains the
+	// _actual_ target URL and the correct credentials to use.
```

Comment:
This description seems to be for what's happening in `registerManagedTransportOptions()`. It may be better to move it there.